### PR TITLE
Disable activation of overlays while in gameplay

### LINKED ISSUE #9743: Add an option to disable "ctr + key actions" in game
**Describe the new feature:** 

As a mania player, I use s-d-f-space-j-k-l for playing, and on many occasions, I have pressed without intention the ctrl key with my left hand while playing, opening direct overlay, (ctr + d) and that has led me to lose several times.
I would like to have an option in the menu to disable ctr + key actions (just as the windows key) to disable these actions in gameplay.

**Proposal designs of the feature:** 

As an option in the main menu where you can choose to disable these actions if they bother you as it does for me.

### PR DESCRIPTION
**Pull request by Game4all**
Closes #9743
Alternative to / supersedes  #9751

# Summary

This adds the ability to disable activation of overlays while in gameplay through a setting that defaults to true. Overlay activation is re-enabled back if the gameplay is paused, enters a break period or the current gameplay is running a replay.

### CODE REVIEW

File: osu.Game.Tests/Visual/Gameplay/TestSceneOverlayActivation.cs
Comment (bdach) [inline osu.Game.Tests/Visual/Gameplay/TestSceneOverlayActivation.cs:17]: let's use proper variable names please

```suggestion
        private OsuConfigManager config { get; set; }
```

(+ usages)

File: osu.Game/Screens/Play/Player.cs
Comment (bdach) [inline osu.Game/Screens/Play/Player.cs:93]: seeing that you're binding to `OsuGame`'s bindable bi-directionally, there's no need to expose this for the test - just resolve `OsuGame` in the test itself, get a bound copy from it and inspect that - it should work the same

File: osu.Game/Screens/Play/Player.cs
Comment (bdach) [inline osu.Game/Screens/Play/Player.cs:219]: this is super awkwardly written to me; if you wanted to centralise logic of changing `OverlayActivationMode` I would suggest to follow the example of `updatePauseOnFocusLostState()` below by extracting the callback to a `updateOverlayActivationMode()` method, and then using it in the callbacks of all these three bindables - this way there's no weird daisy-chaining of value changed callbacks

File: osu.Game/Configuration/OsuConfigManager.cs
Comment (bdach) [inline osu.Game/Configuration/OsuConfigManager.cs:103]: seeing that this is given a default of `true` this tip might need wording adjustment

https://github.com/ppy/osu/blob/fd56db21568a7c71e4390a4b20a69e0922538a88/osu.Game/Screens/Menu/Disclaimer.cs#L193

**Review by bdach — DISMISSED** [review]

File: osu.Game/Screens/Play/Player.cs
Comment (Game4all) [reply osu.Game/Screens/Play/Player.cs:93]: While I agree this should have been the way to do things, this won't work since trying to resolve `OsuGame` in game testing code returns null ( OsuTestBrowser derives from `OsuGameBase`), hence the existence of this protected bindable.

Comment (bdach) [reply osu.Game/Screens/Play/Player.cs:93]: fair enough, would have been good to mention it at the start - this however means that this needs xmldoc

also a tiny bit confused at why this is specifying `Disabled` as its initial value but there is also this

https://github.com/ppy/osu/blob/25ebb8619d2cbaa6fc232ed742d33da91b80d703/osu.Game/Screens/Play/Player.cs#L53

Comment (Game4all) [reply osu.Game/Screens/Play/Player.cs:93]: Bindable initial value should be the same as `InitialOverlayActivationMode` of the screen, I missed that one.

File: osu.Game/Configuration/OsuConfigManager.cs
Comment (Game4all) [reply osu.Game/Configuration/OsuConfigManager.cs:103]: IMO, I don't think the tip needs to be adjusted. I mean, the tip is still half valid in the sense that you can change the skin while playing but you'll have to pause to open the settings and switch skins, but I guess that's what most of players (to not say all) would do.

File: osu.Game/Screens/Play/Player.cs
Comment (bdach) [inline osu.Game/Screens/Play/Player.cs:367]: can't say I like the way this is written too much, feels like too much branching; would probably rewrite to
```csharp
bool canTriggerOverlaysDuringGameplay = DrawableRuleset.IsPaused.Value || !gameplayOverlaysDisabled.Value;
if (DrawableRuleset.HasReplayLoaded.Value || canTriggerOverlaysDuringGameplay)
    OverlayActivationMode.Value = OverlayActivation.UserTriggered;
else
    OverlayActivationMode.Value = OverlayActivation.Disabled;
```

feels more legible that way to me

File: osu.Game/Configuration/OsuConfigManager.cs
Comment (bdach) [reply osu.Game/Configuration/OsuConfigManager.cs:103]: mostly I'd just mention that pausing is necessary to avoid people claiming that the tip is incorrect:

> All settings are dynamic and take effect in real-time. Try pausing and changing the skin while playing!

File: osu.Game.Tests/Visual/Gameplay/TestSceneOverlayActivation.cs
Comment (bdach) [inline osu.Game.Tests/Visual/Gameplay/TestSceneOverlayActivation.cs:12]: would also like a test case for allowing the overlays to be opened by disabling the setting here for completeness

Comment (Game4all) [reply osu.Game.Tests/Visual/Gameplay/TestSceneOverlayActivation.cs:12]: In fact, I had a test to ensure that overlays can be opened when the setting is disabled, but I ended up not committing it as for some reason the setting update doesn't propagate to the loaded test player (whereas this works just fine in game), making the test always fail.

Comment (bdach) [reply osu.Game.Tests/Visual/Gameplay/TestSceneOverlayActivation.cs:12]: That is quite unsettling. I'd like to know why exactly it fails in the test.

Comment (Game4all) [reply osu.Game.Tests/Visual/Gameplay/TestSceneOverlayActivation.cs:12]: For reference the test was:
```csharp
        [Test]
        public void TestGameplayOverlayActivationSettingDisabled()
        {
            AddStep("enable overlay activation during gameplay", () => config.Set(OsuSetting.GameplayDisableOverlayActivation, false));
            AddAssert("activation mode is user triggered", () => testPlayer.OverlayActivationMode == OverlayActivation.UserTriggered);
        }
```
and would fail on the assertion of the overlay activation mode (always returning `OverlayActivationMode.Disabled` whether it should be `OverlayActivationMode.UserTriggered`).
I tried replacing with an `AddUntil` step but I keep getting the same results.

It looks like the setting update doesn't get propagated to the test player.

You can easily replicate by copy pasting into the test cases file.

Comment (bdach) [reply osu.Game.Tests/Visual/Gameplay/TestSceneOverlayActivation.cs:12]: the test doesn't work because it's using the wrong instance of `OsuConfigManager`. `PlayerTestScene` creates and caches its own:

https://github.com/ppy/osu/blob/c09868780e1d03703a18f37e512777fba3afe89c/osu.Game/Tests/Visual/PlayerTestScene.cs#L25-L32

using that one instead of the locally-`[Resolved]` config manager fixes the failure.

on that note I'd also like to say that I find the mentality of discarding the test if it doesn't work correctly quite dangerous. we shouldn't become accustomed to accepting that things just don't work sometimes for unknown reasons. if I hadn't prodded continuously reviewing this, I probably wouldn't have found out that the tests were half-broken due to using the wrong config instance, especially seeing that they *read* sane.

(also not sure how to feel about having to debug this myself, but, alas.)

Comment (Game4all) [reply osu.Game.Tests/Visual/Gameplay/TestSceneOverlayActivation.cs:11]: I apologize for having left the debugging of this to you. I feel really bad for doing it, especially when the cause is that I didn't paid attention enough to the existing code.

**Comment by bdach** [comment]
I've pushed in a menial change that aims to avoid having `OverlayActivation.UserTriggered` specified in two places (now they both refer to a single constant). Unfortunately this can't be done without introducing the constant.

Otherwise this seems ok. In a quick gameplay test it works roughly as expected - the only issue is that the hotkeys make a click sound even if disabled, but that should probably be fixed globally as I imagine it'll be a global issue and is probably out of scope here.

Not approving since I pushed.

**Review by frenzibyte — CHANGES_REQUESTED** [review]
The current logic looks fine to me for now otherwise, but going forward, I would suggest making `OverlayActivationMode` immutable from any component, and letting `OsuGame` get the overlay activation mode of the current active screen or otherwise, this is to avoid changing `OverlayActivationMode` in the wrong time, in this case, while the screen isn't yet pushed as caught below.

---

Also, there's another issue arisen because of this first time using `OverlayActivationMode.Disabled`, toggling the toolbar by `Ctrl+T` displays it frozen: https://streamable.com/rllzwd. 

Given that these small issues currently exist, and that this PR doesn't really need to be merged as soon as possible, maybe better to fix these issues first before merging this.

File: osu.Game/Screens/Menu/Disclaimer.cs
Comment (bdach) [reply osu.Game/Screens/Menu/Disclaimer.cs:193]: that's not really proper English

File: osu.Game/Screens/Play/Player.cs
Comment (bdach) [reply osu.Game/Screens/Play/Player.cs:217]: sounds like we might want to take a lease on the bindable instead of directly binding

File: osu.Game/Screens/Menu/Disclaimer.cs
Comment (frenzibyte) [reply osu.Game/Screens/Menu/Disclaimer.cs:193]: of course, that's what the **maybe** is for, I'm pointing that there's kind of an oxymoron here that shouldn't really exist in such place. (or anywhere)

**Comment by frenzibyte** [comment]
Oh, and also this should allow overlay activation on break times as well, it's a kind of a normal habit for a player to open the chat overlay and talk there in such periods.

File: osu.Game/Screens/Play/Player.cs
Comment (frenzibyte) [reply osu.Game/Screens/Play/Player.cs:217]: oh just realized I didn't mention this, but the settings overlay closing is the result of me setting `Disable overlays during gameplay` on, I'm suggesting to move these binds inside `OnEntering()` where it makes sense to start allowing to change the `OverlayActivationMode`.

Comment (Game4all) [reply osu.Game/Screens/Play/Player.cs:217]: Alternatively, adding a check that `Player` is the current screen should do the trick aswell.

Comment (frenzibyte) [reply osu.Game/Screens/Play/Player.cs:217]: Oh, now that you mention it, this'll also change `OverlayActivationMode` on the results screen, tbh I think moving all this overlay activation mode logic into an `OsuGame` level would be a better choice as I mentioned above https://github.com/ppy/osu/pull/9766#pullrequestreview-463824776 (in a separate PR this would depend on)

Comment (bdach) [reply osu.Game/Screens/Play/Player.cs:217]: taking a lease on the bindable would also work as desired by rolling the changes back, and require no other changes.

Comment (Game4all) [reply osu.Game/Screens/Play/Player.cs:217]: > 
> 
> Oh, now that you mention it, this also changes `OverlayActivationMode` on the results screen, tbh I think moving all this overlay activation mode logic into an `OsuGame` level would be a better choice as I mentioned above [#9766 (review)](https://github.com/ppy/osu/pull/9766#pullrequestreview-463824776) (in a separate PR this would depend on)
> 
> Edit: Adding the `Player` check would fix this as well but it would just make the whole thing look like it shouldn't be there.

Did some debugging as this sounded quite weird to me. Turns out that the the behaviour described in results screen is
because of this: https://github.com/ppy/osu/blob/9a09f97478f063384a03c0d35d23f359cc9d1353/osu.Game/Screens/Ranking/ResultsScreen.cs#L33

image

As seen on the screen, the activation mode is `All`, which is the default activation mode of an `OsuScreen`. Would've been either `UserTriggered` or `Disabled` if  `Player` was to change the `OverlayActivationMode`.

File: osu.Game/Screens/Play/Player.cs
Comment (bdach) [inline osu.Game/Screens/Play/Player.cs:367]: why is this check required?

Comment (Game4all) [reply osu.Game/Screens/Play/Player.cs:367]: This check has been added to prevent `OverlayActivationMode` from being updated when the `Player` isn't the current screen (could be triggered previously by changing the value of the setting while in the player loader, according to this comment: https://github.com/ppy/osu/pull/9766#pullrequestreview-463824776)

Comment (bdach) [reply osu.Game/Screens/Play/Player.cs:367]: if so then I think I would prefer deferring setting up the subscriptions until `OnEntering()` as proposed above.

File: osu.Game.Tests/Visual/Gameplay/TestSceneOverlayActivation.cs
Comment (peppy) [inline osu.Game.Tests/Visual/Gameplay/TestSceneOverlayActivation.cs:25]: why are the tests only covering the disabled case?

File: osu.Game/Screens/Play/Player.cs
Comment (peppy) [inline osu.Game/Screens/Play/Player.cs:59]: exposing this is really confusing. i'd hope it was private.

File: osu.Game/Screens/Play/Player.cs
Comment (peppy) [inline osu.Game/Screens/Play/Player.cs:656]: i'm pretty sure the state is not guaranteed to stop mutating when exiting player. probably needs a call to `updateOverlayActivationMode()` in exiting.

File: osu.Game/Screens/Menu/Disclaimer.cs
Comment (peppy) [reply osu.Game/Screens/Menu/Disclaimer.cs:193]: I don't really think this needed to change.

**Review by peppy — CHANGES_REQUESTED** [review]

**Comment by peppy** [comment]
Honestly, with the amount of complexity here, I wonder if this even needs to be a setting? Is there a use case for setting the setting to false?

File: osu.Game.Tests/Visual/Gameplay/TestSceneOverlayActivation.cs
Comment (bdach) [reply osu.Game.Tests/Visual/Gameplay/TestSceneOverlayActivation.cs:25]: The next test under this one covers disabling the setting.

Comment (peppy) [reply osu.Game.Tests/Visual/Gameplay/TestSceneOverlayActivation.cs:25]: ah, i see. got confused by the `SetUpStep` setting one state. maybe not the best way to structure tests (move the set into each test instead, or via `[TestCase(bool)]`)

File: osu.Game/Screens/Menu/Disclaimer.cs
Comment (bdach) [reply osu.Game/Screens/Menu/Disclaimer.cs:193]: maybe not... I'm just super not looking forward to potentially having someone claim that the tip is wrong because "settings menu can't be opened during gameplay"

File: osu.Game/Screens/Play/Player.cs
Comment (Game4all) [reply osu.Game/Screens/Play/Player.cs:59]: This bindable is marked as protected only for testing purposes, as we can't resolve an `OsuGame` instance through DI in tests. (see this https://github.com/ppy/osu/pull/9766#discussion_r464548375 for reference)

File: osu.Game.Tests/Visual/Gameplay/TestSceneOverlayActivation.cs
Comment (Game4all) [reply osu.Game.Tests/Visual/Gameplay/TestSceneOverlayActivation.cs]: I pushed a commit to make the test steps hopefully clearer

File: osu.Game/Screens/Play/Player.cs
Comment (Game4all) [reply osu.Game/Screens/Play/Player.cs:656]: I'm not sure at all with that one. Isn't the updating of the overlay mode with the change of screens sufficient?

Comment (frenzibyte) [reply osu.Game/Screens/Play/Player.cs:656]: i.e. enabling the "disable overlays during gameplay" setting while in results screen will affect the global value, because `Player` still exists, just suspended.

**Comment by frenzibyte** [comment]
About the first paragraph in https://github.com/ppy/osu/pull/9766#pullrequestreview-463824776, here's what I think of how should `OverlayActivationMode` be mutable (with `OsuGame` itself being responsible for it): https://github.com/frenzibyte/osu/compare/1398234b1bc54fe2e09158ef7aaac929b7209ec8...6285cd4b108985c02461ddf58590745e89d105b7

That would get rid of any unorganized insurances inside `Player` (binding at `OnEntering`, ensuring `OverlayActivationMode` isn't change while not in current screen), would like to hear if it's going the right direction before PRing though.

**Comment by peppy** [comment]
Changing InitialOverlayActivation to just be a screen bindable and is one-way bound with the game one should work amicably.

File: osu.Game/Screens/Play/Player.cs
Comment (Game4all) [reply osu.Game/Screens/Play/Player.cs:656]: ah it's again that case. Guess I'll wait for your proposed changes to get merged and rebase, as it'll fix that case (aswell as the binding on entering)

**Comment by Game4all** [comment]
> 
> Honestly, with the amount of complexity here, I wonder if this even needs to be a setting? Is there a use case for setting the setting to false?

I'm fine with dropping the setting and having this as a default behaviour, as it would also resolve the issue where the overlay activation mode would change because the setting value was updated despite the player being suspended.

**Review by smoogipoo — APPROVED** [review]
Seems pretty fine to me. I'm also not 100% sure on the disclaimer change, but I think the concern pointed out by @bdach is fairly reasonable.

**Pull request merged**
